src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,153 @@
+        {
+            switch (request.TransportCreationParams)
+            {
+                case SharedMemoryTransportCreationParams:
+                    throw new NotImplementedException();
+
+                case NamedPipeTransportCreationParams:
+                    throw new NotImplementedException();
+
+                case TcpTransportCreationParams tcpParams:
+                    request.ConnectionStream = await HandleTcpRequest(tcpParams, ct).ConfigureAwait(false);
+                    break;
+
+                default:
+                    throw new InvalidOperationException("");
+            }
+
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(request, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        private ValueTask<Stream> HandleNamedPipeRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private ValueTask<Stream> HandleSharedMemoryRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private async Task<Stream> HandleTcpRequest(TcpTransportCreationParams request, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            // DNS lookup
+            var ipAddresses = await Dns.GetHostAddressesAsync(request.Hostname, ct).ConfigureAwait(false);
+            if (ipAddresses.Length == 0)
+            {
+                throw new Exception("Hostname did not resolve");
+            }
+
+            // If there is an IP version preference, apply it
+            switch (request.IpAddressPreference)
+            {
+                case SqlConnectionIPAddressPreference.IPv4First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV4);
+                    break;
+
+                case SqlConnectionIPAddressPreference.IPv6First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV6);
+                    break;
+
+                case SqlConnectionIPAddressPreference.UsePlatformDefault:
+                default:
+                    // Not sorting necessary
+                    break;
+            }
+
+            // Attempt to connect to one of the matching IP addresses
+            // @TODO: Handle opening in parallel
+            Socket socket = null;
+            var socketExceptions = new List<Exception>();
+            foreach (var ipAddress in ipAddresses)
+            {
+                var ipEndpoint = new IPEndPoint(ipAddress, request.Port);
+                try
+                {
+                    socket = await OpenSocketAsync(ipEndpoint, ct).ConfigureAwait(false);
+                }
+                catch(Exception e)
+                {
+                    socketExceptions.Add(e);
+                }
+            }
+
+            // If no socket succeeded, throw
+            if (socket is null)
+            {
+                throw new AggregateException(socketExceptions);
```

Comment:
The current logic throws the first SocketException it can find, then an AggregateException if it can't find any. This behaviour change might be intentional though.

Reply:
I'm really lost on how the current error handling is implemented... For this attempt I kinda just winged it. I can change this logic, but I need a bit of guidance. Are you saying it should throw
* First SocketException in the collected exceptions (if there are any)
* AggregateException from all collected exceptions (if there are no socket exceptions in the list)

Reply:
Yes - unless we want to change it. In cases where there are multiple IP addresses which all fail to connect, I can see an argument in favour of returning all of these SocketExceptions.

Reply:
Ok, I'll tweak the error throwing code to do this